src/index.js

```diff
@@ -1,13 +1,34 @@
 // @flow
 
 import type {
+  NavigationAction,
+  NavigationContainer,
   NavigationEventCallback,
   NavigationState,
 } from 'react-navigation';
-import type { Middleware } from 'redux';
+import type { Middleware, Reducer } from 'redux';
 
+import { NavigationActions } from "react-navigation";
 import invariant from 'invariant';
 
+type Navigator = NavigationContainer<*, *, *>;
+type ReducerState = ?NavigationState;
+
+function createNavigationReducer(navigator: Navigator): Reducer<*,*> {
+  const initialState = navigator.router.getStateForAction(
+    NavigationActions.init(),
+    null
+  );
+
+  return (
+    state: ReducerState = initialState,
+    action: NavigationAction
+  ): ReducerState => {
+    const newState = navigator.router.getStateForAction(action, state);
```

Comment:
My only question is whether we should actually return a `null` state if the router returns `null`. I am not sure how `react-navigation` will render a `null` state, but if the router thinks the state should be `null` then I imagine the view code is set up to handle this scenario?

I want to give the view code a read to understand what happens when the state is `null`, but I don't have time this morning. I'll try to get to it over the weekend.